ffi/src/expressions.rs

```diff
@@ -245,3 +245,117 @@ pub extern "C" fn visit_expression_literal_bool(
 ) -> usize {
     wrap_expression(state, Expression::literal(value))
 }
+
+/// Kernel Expression to Engine Expression
+///
+#[repr(C)]
+pub struct EngineExpressionVisitor {
+    /// opaque state pointer
+    pub data: *mut c_void,
+    /// Creates a new field list, optionally reserving capacity up front
+    pub make_expr_list: extern "C" fn(data: *mut c_void, reserve: usize) -> usize,
+
+    /// Visit an `integer` belonging to the list identified by `sibling_list_id`.
+    pub visit_int: extern "C" fn(data: *mut c_void, value: i32) -> usize,
+    pub visit_long: extern "C" fn(data: *mut c_void, value: i64) -> usize,
+    pub visit_short: extern "C" fn(data: *mut c_void, value: i16) -> usize,
+    pub visit_byte: extern "C" fn(data: *mut c_void, value: i8) -> usize,
+    pub visit_float: extern "C" fn(data: *mut c_void, value: f32) -> usize,
+    pub visit_double: extern "C" fn(data: *mut c_void, value: f64) -> usize,
+    pub visit_bool: extern "C" fn(data: *mut c_void, value: bool) -> usize,
+    pub visit_string: extern "C" fn(data: *mut c_void, value: KernelStringSlice) -> usize,
+
+    pub visit_and: extern "C" fn(data: *mut c_void, len: usize) -> usize,
+    pub visit_or: extern "C" fn(data: *mut c_void, len: usize) -> usize,
+    pub visit_variadic_item:
+        extern "C" fn(data: *mut c_void, variadic_id: usize, sub_expr_id: usize),
+    pub visit_not: extern "C" fn(data: *mut c_void, inner_expr: usize) -> usize,
+    pub visit_is_null: extern "C" fn(data: *mut c_void, inner_expr: usize) -> usize,
+
+    pub visit_lt: extern "C" fn(data: *mut c_void, a: usize, b: usize) -> usize,
+    pub visit_le: extern "C" fn(data: *mut c_void, a: usize, b: usize) -> usize,
+    pub visit_gt: extern "C" fn(data: *mut c_void, a: usize, b: usize) -> usize,
+    pub visit_ge: extern "C" fn(data: *mut c_void, a: usize, b: usize) -> usize,
+    pub visit_eq: extern "C" fn(data: *mut c_void, a: usize, b: usize) -> usize,
+
+    pub visit_column: extern "C" fn(data: *mut c_void, name: KernelStringSlice) -> usize,
+}
+
+#[no_mangle]
+pub unsafe extern "C" fn visit_expression(
+    expression: &Expression,
+    visitor: &mut EngineExpressionVisitor,
+) -> usize {
+    fn visit_variadic(
+        visitor: &mut EngineExpressionVisitor,
+        op: &VariadicOperator,
+        exprs: &Vec<Expression>,
+    ) -> usize {
+        let variadic_id = match op {
+            VariadicOperator::And => (visitor.visit_and)(visitor.data, exprs.len()),
+            VariadicOperator::Or => (visitor.visit_or)(visitor.data, exprs.len()),
+        };
```

Comment:
You might be able to do this instead, to reduce redundancy:
```suggestion
        let visit_fn = match op {
            VariadicOperator::And => visitor.visit_and,
            VariadicOperator::Or => visitor.visit_or,
        };
        let variadic_id = visit_fn(visitor.data, exprs.len());
```